Remove the `boot_script` from `experimental_shell_command`/`experimental_run_in_sandbox`

chrisjrn opened the pull request:
This replaces the boot script that is run on startup for every `esc`/`eris` invocation, and replaces it with a reusable rule that creates a digest full of scripts that call `{tool_name} $@`. This digest is then provided as an `immutable_input_digest` and made available on the `PATH` of the `experimental_shell_command`.

While this doesn't do a whole lot in and of itself, it shows how in future you might be able to replace `tools` with a more general `runnable_dependencies` field, though such a field would be subject to significant design work.

chrisjrn commented:
does not block 2.16

kaos commented on src/python/pants/backend/shell/util_rules/adhoc_process_support.py:
```diff
@@ -183,41 +190,68 @@ def _shell_tool_safe_env_name(tool_name: str) -> str:
+@dataclass(frozen=True)
+class BashToolsRequest:
+    tools: tuple[str, ...]
+    rationale: str
```
This looks very much like `BinaryShimsRequest`..

https://github.com/pantsbuild/pants/blob/b587181465f62b483c95b6f18f4c6cafc90033a9/src/python/pants/core/util_rules/system_binaries.py#L172-L224

kaos replied:
But I think you're implementation is superior. Mind fixing the binary shims rule, perhaps? :)

chrisjrn replied:
Well, I'd completely missed `BinaryShims`. That's handy.

What in particular is superior in my implementation? I see, specifically:

* error checking
* using digest operations rather than a bash command to generate the scripts

Is that all?

kaos replied:
all or more. I didn't compare them in depth, only noticed I had used Processes to create symlinks where you used digest operations.

chrisjrn replied:
Great, I'll get that working against this branch, and then throw a separate PR over the fence for you to review.

chrisjrn replied:
See #18184.

kaos replied:
✅

chrisjrn replied:
@kaos I've rebased this branch on top of #18184; if you wish to re-review, you can safely ignore the changes in `system_binaries*.py`

chrisjrn commented on src/python/pants/backend/codegen/protobuf/lint/buf/format_rules.py:
```diff
@@ -1,6 +1,5 @@
 # Copyright 2022 Pants project contributors (see CONTRIBUTORS.md).
```
(reviewed in #18184)

chrisjrn commented on src/python/pants/backend/docker/util_rules/docker_binary.py:
```diff
@@ -148,13 +148,11 @@ async def find_docker(
         BinaryShimsRequest.for_binaries(
```
(from #18184)

chrisjrn commented on src/python/pants/core/util_rules/system_binaries.py:
```diff
@@ -11,7 +11,7 @@
 from dataclasses import dataclass
```
(from #18184)

chrisjrn commented on src/python/pants/core/util_rules/system_binaries_test.py:
```diff
@@ -173,19 +173,16 @@ def test_binary_shims_request(rule_runner: RuleRunner) -> None:
             BinaryShimsRequest.for_binaries(
```
(from #18184)

kaos commented on src/python/pants/backend/shell/util_rules/shell_command_test.py:
```diff
@@ -617,20 +618,9 @@ def test_shell_command_boot_script(rule_runner: RuleRunner) -> None:
-    assert sorted(res.env["TOOLS"].split()) == tools
-    for tool in tools:
-        assert res.env[tool].endswith(f"/{tool.replace('_', '.')}")
+    assert "PATH" in res.env
```
I'd like to preserve the test for the path to `python3.8` here.. (i.e. assert that there is a `python3.8` on the PATH provided.)

kaos reviewed (approved):
🚀

chrisjrn commented:
@kaos per your request, I've added [a test for checking that `PATH` is populated correctly](https://github.com/pantsbuild/pants/pull/18168/files#diff-da19d10a78842b63d5c3e912834bc966bff98ca81a65ec1cadb9a050eb8cc641R598), at which point I discovered that `BASH_BUILTIN_COMMANDS` was overspecified and the new added error checking would break in different ways on different machines, so I've updated builtin.py to include only the functions listed in https://www.gnu.org/software/bash/manual/bash.txt section 4.2 (builtin functions)

kaos commented:
> @kaos per your request, I've added [a test for checking that `PATH` is populated correctly](https://github.com/pantsbuild/pants/pull/18168/files#diff-da19d10a78842b63d5c3e912834bc966bff98ca81a65ec1cadb9a050eb8cc641R598), at which point I discovered that `BASH_BUILTIN_COMMANDS` was overspecified and the new added error checking would break in different ways on different machines, so I've updated builtin.py to include only the functions listed in https://www.gnu.org/software/bash/manual/bash.txt section 4.2 (builtin functions)

I regret not having the source for the original list in the commit/PR. I vaguely recall having extracted that list from `man bash` under the "SHELL BUILTIN COMMANDS"; but, different shells may have slightly different set of builtins.. not sure if it would be worthwhile to accomodate for that though?

However I think we'd want to at least also include the commands listed in section 4.1.

chrisjrn commented:
@kaos 4.1 is all there, along with those from 4.2.

chrisjrn commented:
@kaos I think supporting other shells' builtins would make sense once we start accommodating those shells, but for now we only support `bash`.
